assets/js/components/surveys/CurrentSurvey.js

```diff
@@ -16,32 +16,173 @@
+
+	const { setValues } = useDispatch( CORE_FORMS );
+	const { sendSurveyEvent } = useDispatch( CORE_USER );
+
+	const currentQuestion = questions?.find( ( question ) => {
+		return question.question_ordinal === currentQuestionOrdinal;
+	} );
+
+	const answerQuestion = useCallback( ( answer ) => {
+		// eslint-disable-next-line camelcase
+		if ( currentQuestion?.question_type === 'rating' ) {
+			sendSurveyEvent( 'question_answered', {
+				// eslint-disable-next-line camelcase
+				question_ordinal: currentQuestion?.question_ordinal,
+				answer: {
+					answer_ordinal: answer.answer_ordinal,
+				},
+			} );
+
+			setValues( formName, {
+				answers: [
+					...answers || [],
+					{
+					// eslint-disable-next-line camelcase
+						question_ordinal: currentQuestion?.question_ordinal,
+						answer_ordinal: answer.answer_ordinal,
+					},
+				],
+			} );
+		}
+	}, [ answers, currentQuestion, formName, sendSurveyEvent, setValues ] );
+
+	const ctaOnClick = useCallback( () => {
+		sendSurveyEvent( 'follow_up_link_clicked' );
+		sendSurveyEvent( 'survey_closed' );
+
+		setValues( formName, { hideSurvey: true } );
+	}, [ formName, sendSurveyEvent, setValues ] );
+
+	const dismissSurvey = useCallback( () => {
+		sendSurveyEvent( 'survey_closed' );
 
-	if ( ! usingProxy || ! currentSurvey ) {
+		setValues( formName, { hideSurvey: true } );
+	}, [ formName, sendSurveyEvent, setValues ] );
+
+	// Check to see if a completion trigger has been met.
+	let triggeredCompletion;
+	// We only have trigger conditions for questions that are answered with
+	// ordinal values right now.
+	const ordinalAnswerMap = answers?.length ? answers.reduce( ( acc, answer ) => {
+		return {
+			...acc,
+			[ answer.question_ordinal ]: answer.answer_ordinal,
+		};
+	}, {} ) : [];
+
+	// Use Array.some to avoid looping through all completions; once a matching
+	// completion has been found, treat the survey as complete.
+	completions?.some( ( completion ) => {
+		completion.trigger_condition.some( ( condition ) => {
+			// If a question was answered with the appropriate value, a completion
+			// trigger has been fulfilled and we should treat this survey as complete.
+			if ( condition.answer_ordinal.includes( ordinalAnswerMap[ condition.question_ordinal ] ) ) { // eslint-disable-line camelcase
+				triggeredCompletion = completion;
+				return true;
+			}
+
+			return false;
+		} );
+
+		if ( triggeredCompletion ) {
+			return true;
+		}
+
+		return false;
+	} );
+
+	// If no specific trigger has been found but all questions are answered, use
+	// the first available trigger.
+	if ( ! triggeredCompletion && questions?.length && currentQuestionOrdinal > questions?.length ) {
+		triggeredCompletion = completions[ 0 ];
+	}
+
+	useEffect( () => {
+		if ( triggeredCompletion && ! hasSentCompletionEvent ) {
+			sendSurveyEvent( 'completion_shown', {
+				completion_ordinal: triggeredCompletion.completion_ordinal,
+			} );
+			setHasSentCompletionEvent( true );
```

Comment:
Let's move the set state before sending the actual event above. We're not awaiting it here, but still a good precaution to take.